url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPLinkManager.java

```diff
@@ -0,0 +1,2057 @@
+        /**
+         * The receive buffer.
+         */
+        private ByteBuffer     buffer;
+        /**
+         * The timestamp of the last activity.
+         */
+        private long           lastActivity;
+        /**
+         * Used by TxThread to signal this thread to go directly to the RESET
+         * state (due to a connection failure).
+         */
+        private boolean        resetNeeded;
+
+        /**
+         * Create a new stopped RxThread.
+         */
+        public RxThread() {
+            super();
+
+            currentState = RxThreadStates.UNINITIALISED;
+            shutdownThread = false;
+            myThread = null;
+            socketSelector = null;
+            lastActivity = System.currentTimeMillis();
+            resetNeeded = false;
+        }
+
+        @Override
+        public void run() {
+            log.debug("RxThread: thread starting");
+            watchdog.start();
+            myThread = Thread.currentThread();
+            boolean moreToDo = false;
+            lastActivity = System.currentTimeMillis();
+
+            buffer = ByteBuffer.allocate(LegacyOCPMessage.OCP_MAX_LENGTH);
+            buffer.order(ByteOrder.BIG_ENDIAN); // network order
+
+            while (true) {
+                watchdog.pat();
+                if (resetNeeded) {
+                    log.info("RxThread: socket shutdown detected");
+                    currentState = RxThreadStates.RESET;
+                }
+                moreToDo = internalRun();
+
+                if (!moreToDo) {
+                    try {
+                        socketSelector.select(1 * MILLISECONDS);
+                        socketSelector.selectedKeys().clear();
+                    } catch (IOException e) {
+                        log.error("RxThread: exception while waiting on "
+                                + "selector", e);
+                        currentState = RxThreadStates.RESET;
+                    }
+                }
+
+                if (System.currentTimeMillis() - lastActivity > timeout) {
+                    // Timeout: reset state machine and try again
+                    log.warn("RxThread: timeout detected while in state "
+                            + currentState.name());
+                    currentState = RxThreadStates.RESET;
+                }
+
+                if (shutdownThread) {
+                    cleanup();
+                    break;
+                }
+            }
+            watchdog.stop();
+            log.debug("RxThread: thread exiting");
+        }
+
+        /**
+         * Shut down this thread.
+         */
+        public void shutdownThread() {
+            log.debug("RxThread: thread shutdown requested");
+            shutdownThread = true;
+            if (myThread != null) {
+                while (myThread.isAlive()) {
+                    try {
+                        myThread.join();
+                    } catch (InterruptedException e) { // CSIGNORE: EmptyBlock
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)